File: src/regent-cli/src/features/chat/domain/questionnaire.ts
// MIRROR of src/crates/regent-kernel/src/contracts/questionnaire.rs — the
// structured-question contract. Hand-copied because the CLI and the Desktop app
// are separate build roots that cannot import Rust or each other; both copies
// are byte-identical below the header and checked against the Rust by
// scripts/tests/verify-questionnaire-schema.py in the `parity` CI job.
// Edit the Rust first, then both copies, or CI fails.
//
// Pure: types plus the stepper helpers every surface needs. No I/O, no UI.

export type QuestionKind = "single_select" | "multi_select" | "text" | "confirm" | "rank";

export interface QuestionOption {
  readonly id: string;
  readonly label: string;
  readonly description?: string;
}

export interface Question {
  readonly id: string;
  readonly prompt: string;
  /** Very short chip label beside the question ("Auth method", "Scope"). */
  readonly header?: string;
  readonly kind: QuestionKind;
  readonly options?: readonly QuestionOption[];
  /** "Something else" free-text row. Defaults to true on the wire. */
  readonly allow_custom?: boolean;
  readonly required?: boolean;
}

export interface Questionnaire {
  readonly id: string;
  readonly questions: readonly Question[];
}

export type Answer =
  /** Option ids in the order chosen — which is also the ranking for `rank`. */
  | { readonly kind: "selected"; readonly option_ids: readonly string[] }
  | { readonly kind: "text"; readonly text: string }
  | { readonly kind: "confirmed"; readonly yes: boolean }
  | { readonly kind: "skipped" };

export interface QuestionnaireAnswer {
  readonly questionnaire_id: string;
  /** One entry per answered question, keyed by `Question.id`. */
  readonly answers: readonly (readonly [string, Answer])[];
  readonly cancelled?: boolean;
}

/** True when the kind is answered by picking from `options`. */
export function needsOptions(kind: QuestionKind): boolean {
  return kind === "single_select" || kind === "multi_select" || kind === "rank";
}

/** True when more than one option may be chosen. */
export function isMulti(kind: QuestionKind): boolean {
  return kind === "multi_select" || kind === "rank";
}

/** Options with a guaranteed array, so callers never re-check for undefined. */
export function optionsOf(question: Question): readonly QuestionOption[] {
  return question.options ?? [];
}

/** The free-text row is on unless the questionnaire explicitly turns it off. */
export function allowsCustom(question: Question): boolean {
  return question.allow_custom !== false;
}

/** Index of the first question with no answer yet, or -1 when all are done. */
export function nextUnanswered(
  questionnaire: Questionnaire,
  answers: QuestionnaireAnswer["answers"],
): number {
  const answered = new Set(answers.map(([id]) => id));
  return questionnaire.questions.findIndex((q) => !answered.has(q.id));
}

/** Records one answer, replacing any earlier answer for the same question. */
export function applyAnswer(
  answers: QuestionnaireAnswer["answers"],
  questionId: string,
  answer: Answer,
): QuestionnaireAnswer["answers"] {
  const kept = answers.filter(([id]) => id !== questionId);
  return [...kept, [questionId, answer] as const];
}

/** The answer recorded for `questionId`, if any. */
export function answerFor(
  answers: QuestionnaireAnswer["answers"],
  questionId: string,
): Answer | undefined {
  return answers.find(([id]) => id === questionId)?.[1];
}

/** Human-readable summary of one answer, for a transcript line or a chip. */
export function describeAnswer(question: Question, answer: Answer): string {
  switch (answer.kind) {
    case "selected":
      return answer.option_ids
        .map((id) => optionsOf(question).find((o) => o.id === id)?.label ?? id)
        .join(", ");
    case "text":
      return answer.text;
    case "confirmed":
      return answer.yes ? "yes" : "no";
    case "skipped":
      return "(skipped)";
  }
}

/**
 * Maps one free-text reply onto a typed answer — the non-TTY / piped path, and
 * the mirror of `parse_text_answer` in the Rust. `2`, `1,3`, `yes`/`no`, or
 * anything else as custom text. Ambiguity stays free text rather than silently
 * meaning an option.
 */
export function parseTextAnswer(question: Question, reply: string): Answer {
  const text = reply.trim();
  if (text === "") return { kind: "skipped" };
  if (question.kind === "confirm") {
    const yes = affirmative(text);
    return yes === undefined ? { kind: "text", text } : { kind: "confirmed", yes };
  }
  if (needsOptions(question.kind)) {
    const ids = parseIndices(question, text);
    if (ids) return { kind: "selected", option_ids: ids };
  }
  return { kind: "text", text };
}

function parseIndices(question: Question, text: string): string[] | undefined {
  const tokens = text
    .split(/[,;\s]+/)
    .map((t) => t.trim())
    .filter((t) => t !== "");
  if (tokens.length === 0) return undefined;
  if (tokens.length > 1 && !isMulti(question.kind)) return undefined;
  const options = optionsOf(question);
  const ids: string[] = [];
  for (const token of tokens) {
    if (!/^\d+$/.test(token)) return undefined;
    const option = options[Number(token) - 1];
    if (!option) return undefined;
    if (!ids.includes(option.id)) ids.push(option.id);
  }
  return ids;
}

const YES = new Set(["y", "yes", "yeah", "yep", "ok", "okay", "sure", "1", "true"]);
const NO = new Set(["n", "no", "nope", "nah", "2", "false"]);

function affirmative(text: string): boolean | undefined {
  const word = text.trim().toLowerCase();
  if (YES.has(word)) return true;
  if (NO.has(word)) return false;
  return undefined;
}

File: src/crates/regent-kernel/src/contracts/questionnaire.rs
//! The structured-question contract: one typed object an agent can put in
//! front of a human, and one typed object that comes back. Lives in the kernel
//! because regent-tools authors it, regent-deacon and regent-gateway ferry it,
//! and every surface renders it — the kernel is the only crate all of them
//! already depend on.
//!
//! Deliberately batch-shaped: a questionnaire carries ALL its questions, and a
//! surface answers them all in one reply. That keeps the existing single-slot
//! pending-approval channel correct as-is — "1 of 3" is a client-side stepper
//! over one request, not three round-trips.

use serde::{Deserialize, Serialize};

/// Caps are tight on purpose. A card with nine options is a menu: it does not
/// fit a Telegram inline keyboard or a terminal window, and it means the model
/// should have asked a different question.
pub const MAX_QUESTIONS: usize = 5;
pub const MAX_OPTIONS: usize = 6;
pub const MIN_OPTIONS: usize = 2;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum QuestionKind {
    SingleSelect,
    MultiSelect,
    Text,
    Confirm,
    /// Selection order is the ranking, so `Answer::Selected` covers it too.
    Rank,
}

impl QuestionKind {
    /// True when the kind is answered by picking from `options`.
    #[must_use]
    pub fn needs_options(self) -> bool {
        matches!(self, Self::SingleSelect | Self::MultiSelect | Self::Rank)
    }

    /// True when more than one option may be chosen.
    #[must_use]
    pub fn is_multi(self) -> bool {
        matches!(self, Self::MultiSelect | Self::Rank)
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct QuestionOption {
    pub id: String,
    pub label: String,
    /// The line under the label that carries the actual decision content.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Question {
    pub id: String,
    pub prompt: String,
    /// Very short chip label beside the question ("Auth method", "Scope") —
    /// what makes the card read as a form field rather than a paragraph.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub header: Option<String>,
    pub kind: QuestionKind,
    #[serde(default)]
    pub options: Vec<QuestionOption>,
    /// Offer a "Something else" free-text row beside the options. Defaults to
    /// TRUE and the tool tells the model never to author its own "Other", so a
    /// model never burns one of its few option slots re-inventing the escape
    /// hatch.
    #[serde(default = "yes")]
    pub allow_custom: bool,
    /// A skippable question resolves to `Skipped`; `required` marks the ones a
    /// surface should not let the user step past.
    #[serde(default)]
    pub required: bool,
}

const fn yes() -> bool {
    true
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Questionnaire {
    pub id: String,
    pub questions: Vec<Question>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum Answer {
    /// Option ids in the order the user chose them — which is also the ranking
    /// for `Rank`, so one variant covers select and rank.
    Selected {
        option_ids: Vec<String>,
    },
    Text {
        text: String,
    },
    Confirmed {
        yes: bool,
    },
    Skipped,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct QuestionnaireAnswer {
    pub questionnaire_id: String,
    /// One entry per answered question, keyed by `Question.id`.
    pub answers: Vec<(String, Answer)>,
    /// True when the user dismissed the whole card rather than answering.
    #[serde(default)]
    pub cancelled: bool,
}

impl QuestionnaireAnswer {
    /// The answer for `question_id`, if the user gave one.
    #[must_use]
    pub fn get(&self, question_id: &str) -> Option<&Answer> {
        self.answers
            .iter()
            .find(|(id, _)| id == question_id)
            .map(|(_, answer)| answer)
    }
}

/// Rejects a questionnaire the surfaces could not render, before it leaves the
/// tool — the model gets the violation back and corrects itself rather than a
/// human getting a broken card.
///
/// # Errors
/// Returns the first violation found, phrased for the model.
pub fn validate(q: &Questionnaire) -> Result<(), String> {
    if q.id.trim().is_empty() {
        return Err("questionnaire needs a non-empty id".to_owned());
    }
    if q.questions.is_empty() || q.questions.len() > MAX_QUESTIONS {
        return Err(format!(
            "a questionnaire needs 1..={MAX_QUESTIONS} questions, got {}",
            q.questions.len()
        ));
    }
    let mut seen = Vec::with_capacity(q.questions.len());
    for question in &q.questions {
        validate_question(question)?;
        if seen.contains(&question.id.as_str()) {
            return Err(format!("duplicate question id {:?}", question.id));
        }
        seen.push(question.id.as_str());
    }
    Ok(())
}

fn validate_question(question: &Question) -> Result<(), String> {
    if question.id.trim().is_empty() {
        return Err("every question needs a non-empty id".to_owned());
    }
    if question.prompt.trim().is_empty() {
        return Err(format!("question {:?} needs a prompt", question.id));
    }
    let count = question.options.len();
    if question.kind.needs_options() {
        if !(MIN_OPTIONS..=MAX_OPTIONS).contains(&count) {
            return Err(format!(
                "question {:?} is a {:?} and needs {MIN_OPTIONS}..={MAX_OPTIONS} options, got {count}",
                question.id, question.kind
            ));
        }
    } else if count > 0 {
        return Err(format!(
            "question {:?} is a {:?} and must not carry options",
            question.id, question.kind
        ));
    }
    let mut seen = Vec::with_capacity(count);
    for option in &question.options {
        if option.id.trim().is_empty() || option.label.trim().is_empty() {
            return Err(format!(
                "every option of question {:?} needs an id and a label",
                question.id
            ));
        }
        if seen.contains(&option.id.as_str()) {
            return Err(format!(
                "duplicate option id {:?} in question {:?}",
                option.id, question.id
            ));
        }
        seen.push(option.id.as_str());
    }
    Ok(())
}

/// Renders a questionnaire as numbered plain text — the universal fallback for
/// a surface that cannot draw a card (an old client, WeChat, a piped stdin).
/// Whatever renders this must also be able to parse [`parse_text_answer`].
#[must_use]
pub fn render_text(q: &Questionnaire) -> String {
    let mut out = String::new();
    let many = q.questions.len() > 1;
    for (index, question) in q.questions.iter().enumerate() {
        if many {
            out.push_str(&format!("({} of {}) ", index + 1, q.questions.len()));
        }
        out.push_str(&question.prompt);
        out.push('\n');
        match question.kind {
            QuestionKind::Confirm => out.push_str("  reply yes or no\n"),
            QuestionKind::Text => out.push_str("  reply with your answer\n"),
            kind => {
                for (n, option) in question.options.iter().enumerate() {
                    out.push_str(&format!("  {}. {}", n + 1, option.label));
                    if let Some(description) = &option.description {
                        out.push_str(&format!(" — {description}"));
                    }
                    out.push('\n');
                }
                out.push_str(if kind.is_multi() {
                    "  reply with numbers (e.g. 1,3) or your own answer\n"
                } else {
                    "  reply with a number or your own answer\n"
                });
            }
        }
    }
    out
}

/// Maps one whole text reply onto the questionnaire: one line per question, in
/// the order they were rendered. Fewer lines than questions leaves the rest
/// `Skipped` — which is honest, rather than guessing that one sentence answered
/// three questions. This is the text surfaces' inverse of [`render_text`].
#[must_use]
pub fn parse_text_reply(q: &Questionnaire, reply: &str) -> Vec<(String, Answer)> {
    let mut lines = reply.lines().map(str::trim).filter(|l| !l.is_empty());
    q.questions
        .iter()
        .map(|question| {
            let answer = lines
                .next()
                .map_or(Answer::Skipped, |line| parse_text_answer(question, line));
            (question.id.clone(), answer)
        })
        .collect()
}

/// Maps one free-text reply onto a typed answer for `question`: `2`, `1,3`,
/// `yes`/`no`, or anything else as custom text. The inverse of the numbered
/// rendering above, and the single parser every text surface shares.
#[must_use]
pub fn parse_text_answer(question: &Question, reply: &str) -> Answer {
    let text = reply.trim();
    if text.is_empty() {
        return Answer::Skipped;
    }
    if question.kind == QuestionKind::Confirm {
        return match affirmative(text) {
            Some(yes) => Answer::Confirmed { yes },
            None => Answer::Text {
                text: text.to_owned(),
            },
        };
    }
    if question.kind.needs_options()
        && let Some(ids) = parse_indices(question, text)
    {
        return Answer::Selected { option_ids: ids };
    }
    Answer::Text {
        text: text.to_owned(),
    }
}

/// `1` / `1,3` / `2 4` → option ids, in the order given (which is the ranking).
/// `None` when any token is not a valid 1-based option number, so a reply like
/// "3 or 4, whichever" stays free text instead of silently meaning "3".
fn parse_indices(question: &Question, text: &str) -> Option<Vec<String>> {
    let tokens: Vec<&str> = text
        .split([',', ' ', ';'])
        .map(str::trim)
        .filter(|t| !t.is_empty())
        .collect();
    if tokens.is_empty() || (tokens.len() > 1 && !question.kind.is_multi()) {
        return None;
    }
    let mut ids = Vec::with_capacity(tokens.len());
    for token in tokens {
        let index: usize = token.parse().ok()?;
        let option = question.options.get(index.checked_sub(1)?)?;
        if !ids.contains(&option.id) {
            ids.push(option.id.clone());
        }
    }
    Some(ids)
}

/// yes/no in the words people actually type. `None` = neither.
fn affirmative(text: &str) -> Option<bool> {
    match text.trim().to_ascii_lowercase().as_str() {
        "y" | "yes" | "yeah" | "yep" | "ok" | "okay" | "sure" | "1" | "true" => Some(true),
        "n" | "no" | "nope" | "nah" | "2" | "false" => Some(false),
        _ => None,
    }
}

/// Human-readable summary of one answer — what the tool hands back to the
/// model and what a transcript shows after the card is answered.
#[must_use]
pub fn describe_answer(question: &Question, answer: &Answer) -> String {
    match answer {
        Answer::Selected { option_ids } => option_ids
            .iter()
            .map(|id| label_of(question, id))
            .collect::<Vec<_>>()
            .join(", "),
        Answer::Text { text } => text.clone(),
        Answer::Confirmed { yes } => if *yes { "yes" } else { "no" }.to_owned(),
        Answer::Skipped => "(skipped)".to_owned(),
    }
}

fn label_of(question: &Question, option_id: &str) -> String {
    question
        .options
        .iter()
        .find(|option| option.id == option_id)
        .map_or_else(|| option_id.to_owned(), |option| option.label.clone())
}

#[cfg(test)]
#[path = "questionnaire_tests.rs"]
mod tests;
